pkg/build/stage/image_spec_mutate.go

```diff
@@ -0,0 +1,214 @@
+package stage
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/werf/common-go/pkg/util"
+	"github.com/werf/werf/v2/pkg/config"
+	"github.com/werf/werf/v2/pkg/container_backend"
+	"github.com/werf/werf/v2/pkg/image"
+	"github.com/werf/werf/v2/pkg/storage"
+)
+
+var ErrImageSpecMutateStageIsNotSupported = fmt.Errorf("error creating image spec mutate stage: can't mutate local image. please provide repo with --repo flag or WERF_REPO env variable")
+
+type ImageSpecMutateStage struct {
+	*BaseStage
+	imageSpecConfig *config.ImageSpec
+}
+
+func GenerateImageSpecMutateStage(imageSpec *config.ImageSpec, baseStageOptions *BaseStageOptions, stagesStorage storage.PrimaryStagesStorage) (*ImageSpecMutateStage, error) {
+	return newImageSpecMutateStage(imageSpec, baseStageOptions, stagesStorage)
+}
+
+func newImageSpecMutateStage(imageSpec *config.ImageSpec, baseStageOptions *BaseStageOptions, stagesStorage storage.PrimaryStagesStorage) (*ImageSpecMutateStage, error) {
+	s := &ImageSpecMutateStage{}
+	switch stagesStorage.(type) {
```

Comment:
There’s no need to check the stage storage here, as validation is already performed at this point: https://github.com/werf/werf/blob/20822ace60305f44a2c522325a9a512a08d6100e/pkg/build/build_phase.go#L1430-L1435. If we leave the validation only there, the user will be able to use these stages, but not build them.